FileUtility.java

```diff
@@ -0,0 +1,92 @@
+import java.io.*;
+import java.text.SimpleDateFormat;
+
+import javax.swing.*;
+import java.util.*;
+import javax.swing.filechooser.FileNameExtensionFilter;
+import org.apache.poi.ss.usermodel.*;
+import org.apache.poi.xssf.usermodel.*;
+
+public class FileUtility extends ReDoReport {
+	
+	 static JFileChooser fileChooser;
+	 	static ArrayList<String> projects = new ArrayList<>();
+		static ArrayList<Date> dates = new ArrayList<>();
+		static ArrayList<String> names = new ArrayList<>();
+		static ArrayList<Double> duration = new ArrayList<>();
+	 public static void findFile() {
+		fileChooser = new JFileChooser(".");
+		FileNameExtensionFilter filter = new FileNameExtensionFilter("Excel File","xlsx");
+		fileChooser.setCurrentDirectory(new File(System.getProperty("user.home")));
+		//filter that will only show Excel files
+		fileChooser.setFileFilter(filter);
+		fileChooser.setDialogTitle("Select Input File");
+		
+		int status = fileChooser.showDialog(fileChooser, "Open");
+		try { //options option to read file and executes if yes
+			if(status == JFileChooser.APPROVE_OPTION) {
+				 fileChooser.getSelectedFile();
+				
+				int n = JOptionPane.showConfirmDialog(fileChooser, "Read File?", "File Reader", 
+						JOptionPane.YES_NO_OPTION);
+				if(n == JOptionPane.OK_OPTION) { 
+					readFile(fileChooser);
+				}else if (n == JOptionPane.NO_OPTION) {
+					findFile();
+				}
+		}
+		} catch (Exception e) {
+			e.printStackTrace();
+		}
+	}
+	public static void readFile(JFileChooser fileChooser) throws IOException{
+		
+		File selectedFile = fileChooser.getSelectedFile();
+		FileInputStream fInput = new FileInputStream(selectedFile);
+		XSSFWorkbook wb = new XSSFWorkbook(fInput);
+		XSSFSheet sheet0 = wb.getSheetAt(1); // <-----change to 0 when finished with tool
+		Iterator<Row> rowIt = sheet0.rowIterator();
```

Comment:
Another preface that I'm not familiar with the libraries your using, or the use case of this method, so bear with me...

It looks like we're using iterators to home in on columns cell. We grab the row iterator, traverse it to the end, then grab cell iterators to traverse down a column. When used in this manner, iterators are like a linked-list; for every row _N_ in the sheet, we have to call `rowIt.next()` that many times. Then, we do the same things for the columns. If the rows and columns are really big, this might become a bottleneck (I've no clue about how big this application will be handling, so this is another caveat).

In my head, a spreadsheet knows how many rows and columns it has, so we could jump directly to the row / column of interest without traversing them all. Does the API / library provide this functionality? That is, rather than using iterators, can we randomly access a row / column? That could speed us up, particularly when finding the final row in the sheet.

Reply:
Maybe this is a start? https://poi.apache.org/apidocs/org/apache/poi/xssf/usermodel/XSSFSheet.html#getLastRowNum--

Reply:
there is a way to get specific rows and columns (treat the sheet like a 2D array) however I'm iterating so it puts each cell of type "x" into a corresponding array which then I need to figure how to manipulate/fill empty rows so a pivot table can be created/and perform various calculations needed